include/xrpl/protocol/STAmount.h

```diff
@@ -684,6 +684,48 @@ isXRP(STAmount const& amount)
     return amount.native();
 }
 
+/** returns true if adding or subtracting results in less than or equal to
+ * 0.01% precision loss **/
+inline bool
+isAddable(STAmount const& amt1, STAmount const& amt2)
+{
+    // special case: adding anything to zero is always fine
+    if (amt1 == beast::zero || amt2 == beast::zero)
+        return true;
+
+    // special case: adding two xrp amounts together.
+    // this is just an overflow check
+    if (isXRP(amt1) && isXRP(amt2))
+    {
+        XRPAmount A = (amt1.signum() == -1 ? -(amt1.xrp()) : amt1.xrp());
+        XRPAmount B = (amt2.signum() == -1 ? -(amt2.xrp()) : amt2.xrp());
+
+        XRPAmount finalAmt = A + B;
```

Comment:
As underlying type of the XRPAmount  is signed, this is incorrect (undefined behavior in case of the overflow).
Consider something like

```
 if( amt1.signum() != amt2.signum()) return true; 
 std::int64_t const A = amt1.xrp().drops();
 std::int64_t const B = (std::int64_t)(((std::uint64_t)A) + ((std::uint64_t)amt2.xrp().drops()));
 return  ((A >= 0) && (B >= A)) || ( (A<0) && B <= A));
```

Reply:
potential fix: https://github.com/XRPLF/rippled/pull/5185/commits/9b989388b2cc4fb0a299ecd6c087a658ae7d7e7a